### PR TITLE
feat: Allow URL encoded dataset in libhoney endpoint paths

### PR DESCRIPTION
**Pull request by MikeGoldsmith**
## Which problem is this PR solving?

Allow URL encoded dataset names in /event and /batch endpoint.

For example, to send data to a dataset with the name `foo/bar`, you can now send data like this:

```
curl -i -H "x-honeycomb-team: <api-key>" -X POST 'http://localhost:8080/1/events/foo%2Fbar' -d '{"name":"mike"}'
```

The dataset name `foo%2Fbar` is URL decoded into `foo/bar` and used as the dataset name. When forwarding events to the Honeycomb API, the dataset name is URL encoded again to ensure it's received into HNY correctly.

## Short description of the changes
- Update mux router to allow encoded paths
- Add utility function to get the dataset from mux vars and url escape the value
- Add tests for utility function
- Simplify batch event processing by removing `batchedEventToEvent`, adding `getSampleRate` batchEvent member func and re-using apiHost and dataset

### CODE REVIEW

**Comment by MikeGoldsmith** [comment]
Waiting on the following PR:
- https://github.com/honeycombio/refinery/pull/1200

File: route/route.go [hunk 1]
```diff
@@ -969,3 +974,16 @@ func (r *Router) AddOTLPMuxxer(muxxer *mux.Router) {
...
+func getDatasetFromRequest(req *http.Request) (string, error) {
+	dataset := mux.Vars(req)["datasetName"]
+	if dataset == "" {
+		return "", fmt.Errorf("missing dataset name")
+	}
+	dataset, err := url.PathUnescape(dataset)
+	if err != nil {
+		return "", err
+	}
+	fmt.Println("dataset", dataset)
```
Comment (VinozzZ) [inline route/route.go:987]: Is this print line intended?

Comment (MikeGoldsmith) [reply route/route.go:987]: Nope, it can be removed. 

'''suggestion
'''

**Review by VinozzZ — APPROVED** [review]

File: route/route_test.go [hunk 2]
```diff
@@ -612,3 +613,109 @@ func TestGRPCHealthProbeWatch(t *testing.T) {
...
+			datasetName:         "foo%5Fbar",
+			expectedDatasetName: "foo_bar",
+		},
+		{
+			name:                "dataset name with URL encoded tilde",
+			datasetName:         "foo%7Ebar",
+			expectedDatasetName: "foo~bar",
+		},
+		{
+			name:                "dataset name with URL encoded period",
```
Comment (kentquirk) [inline route/route_test.go:680]: just for completeness, how about a dataset name with a `%` symbol?

File: route/route.go [hunk 3]
```diff
@@ -678,8 +681,10 @@ func (r *Router) batchedEventToEvent(req *http.Request, bev batchedEvent, apiKey
 	eventTime := bev.getEventTime()
 	// TODO move the following 3 lines outside of this loop; they could be done
 	// once for the entire batch instead of in every event.
-	vars := mux.Vars(req)
```
Comment (kentquirk) [inline route/route.go:681]: Is this the time where we actually implement this TODO given that we're messing with this code?

**Review by kentquirk — APPROVED** [review]
I don't need to block it, but given that there's an old TODO there and this code is being touched...

File: route/route_test.go (hunk 2 above)
Comment (MikeGoldsmith) [reply route/route_test.go:680]: Added extra tests in https://github.com/honeycombio/refinery/pull/1199/commits/492e57250ae9b58d5e18d64c93ac24187755bf3b

File: route/route.go (hunk 3 above)
Comment (MikeGoldsmith) [reply route/route.go:681]: Moved apihost and dataset lookup outside of the loop in https://github.com/honeycombio/refinery/pull/1199/commits/458d27f3c97ddbfa44c59771a4ea17c875bfafd0

**Pull request merged**